Conch/lib/Test/Conch/Validation.pm

```diff
@@ -0,0 +1,285 @@
+
+=item C<description>
+
+Optional description of the test case. Provides documentation and adds the
+description to test failure messages.
+
+=item C<debug>
+
+Optional boolean flag to provide additional diagnostic information when running
+the case using L<Test::More::diag>. This is helpful during development of test
+cases, but should be removed before committing.
+
+=back
+
+Example:
+
+	test_validation(
+		'Conch::Validation::TestValidation',
+		cases => [
+			{
+				data        => { hello => 'world' },
+				dies        => 1,
+				success_num => 3,
+				failure_num => 3,
+				description => 'Hello world test case',
+				debug       => 1
+			}
+		]
+	);
+
+
+=cut
+
+sub test_validation {
+	my $validation_module = shift;
+	my %args              = @_;
+
+	use_ok($validation_module)
+		|| diag "$validation_module fails to compile" && return;
+
+	my $device = Conch::Model::Device->new( $args{device})
+		if $args{device};
+
+	my $hw_product_profile = Conch::Class::HardwareProductProfile->new(
+		$args{hardware_product}->{profile}->%* )
+		if $args{hardware_product} && $args{hardware_product}->{profile};
+	my $hw_product =
+		Conch::Class::HardwareProduct->new( $args{hardware_product}->%*,
+		profile => $hw_product_profile );
+
+	my $rack = Conch::Class::DatacenterRack->new(
+		$args{device_location}->{datacenter_rack}->%* )
+		if $args{device_location} && $args{device_location}->{datacenter_rack};
+	my $device_location =
+		Conch::Class::DeviceLocation->new( $args{device_location}->%*,
+		datacenter_rack => $rack );
+
+	my $validation = $validation_module->new(
+		device           => $device,
+		device_location  => $device_location,
+		device_settings  => $args{device_settings} || {},
+		hardware_product => $hw_product,
+	);
+	isa_ok( $validation, $validation_module )
+		|| diag "$validation_module->new failed" && return;
+
+	ok( defined( $validation->name ) )
+		|| diag("'name' attribute must be defined for $validation_module");
+	ok( defined( $validation->version ) )
+		|| diag("'version' attribute must be defined for $validation_module");
+	ok( defined( $validation->category ) )
+		|| diag("'category' attribute should be defined for $validation_module");
+	ok( defined( $validation->description ) )
+		|| diag("'description' attribute should be defined for $validation_module");
+
+	for my $case_index ( 0 .. $args{cases}->$#* ) {
+		my $case        = $args{cases}->[$case_index];
+		my $data        = $case->{data} || {};
+		my $pretty_data = substr( np($data), 2 );
+		my $debug       = $case->{debug};
+
+		my $msg_prefix = 'Case #' . ( $case_index + 1 );
+		$msg_prefix .=
+			$case->{description} ? ' [' . $case->{description} . ']:' : ':';
+
+		diag("$msg_prefix input data: $pretty_data") if $debug;
+
+		$validation->clear_results;
+
+		if ( $case->{dies} ) {
+			diag("$msg_prefix should die") if $debug;
+			dies_ok( sub { $validation->run_unsafe($data) } )
+				|| diag "$msg_prefix Was expecting $validation_module "
```

Comment:
why the diag here? `dies_ok` accepts a message parameter. https://metacpan.org/pod/Test::Exception#dies_ok

Reply:
I'd prefer that we use the native syntax, since the only reason for variation is a formatting nit. Test names show up in the TAP reports when failures occur and are functionally different than diagnostic output.